ros/src/waypoint_updater/waypoint_updater.py

```diff
@@ -33,20 +35,67 @@ def __init__(self):
 
         # TODO: Add a subscriber for /traffic_waypoint and /obstacle_waypoint below
 
-
         self.final_waypoints_pub = rospy.Publisher('final_waypoints', Lane, queue_size=1)
 
         # TODO: Add other member variables you need below
+        self.base_waypoints = None
+        self.last_ts = None
 
         rospy.spin()
 
     def pose_cb(self, msg):
-        # TODO: Implement
-        pass
+        # stay still until base_waypoints is defined
+        if self.base_waypoints is None:
+            return
+        rospy.loginfo(rospy.get_caller_id() + ": current pose received")
+        # debug
+        print("pose_cb message: ", msg)
+
+        current_ts = msg.header.stamp.to_sec()
+        # debug
+        print("ts: ", current_ts)
+        current_pos = msg.pose.position # x, y, z cartesian coordinate
+        current_ori = msg.pose.orientation # x, y, z, w quaternion
+        (roll, pitch, yaw) = euler_from_quaternion ([current_ori.x,current_ori.y,current_ori.z,current_ori.w])
+        # debug
+        print("roll, pitch, yaw: ", roll, pitch, yaw)
+
+        # publish dummy
+        if self.last_ts is not None:
+            dt = current_ts - self.last_ts
+            print("dt: ", dt)
+            v = 1
+            v_x = v * numpy.cos(yaw)
+            v_y = v * numpy.sin(yaw)
+
+            final_waypoints = Lane()
+            final_waypoints.header = msg.header
+            for i in range(LOOKAHEAD_WPS):
+                dummy_pose = PoseStamped()
+                dummy_pose.pose.position.x = msg.pose.position.x + i * v_x * dt
+                dummy_pose.pose.position.y = msg.pose.position.y + i * v_y * dt
+                dummy_pose.pose.position.z = msg.pose.position.z
+                dummy_pose.pose.orientation.x = msg.pose.orientation.x
+                dummy_pose.pose.orientation.y = msg.pose.orientation.y
+                dummy_pose.pose.orientation.z = msg.pose.orientation.z
+                dummy_pose.pose.orientation.w = msg.pose.orientation.w
+
+                dummy_waypoint = Waypoint()
+                dummy_waypoint.pose = dummy_pose
+                final_waypoints.waypoints.append(dummy_waypoint)
+            # print(final_waypoints)
+            self.final_waypoints_pub.publish(final_waypoints)
+            rospy.loginfo(rospy.get_caller_id() + ": {} final waypoints published".format(len(final_waypoints.waypoints)))
```

Comment:
I don't now what style you want to use, but I never exceed the 80 th column. I find it easier to read through.